From d7368541ddaa3f1c6e0832555d08e7917301d30d Mon Sep 17 00:00:00 2001
From: Alvin Wong <alvin@alvinhc.com>
Date: Sat, 22 Apr 2023 22:07:06 +0300
Subject: [PATCH 3/7] [asan][test][win] Set __COMPAT_LAYER=RunAsInvoker when
 running tests

The `TestCases/asan_update_allocation.cpp` test on i686 MinGW target
happens to trigger a heuristic on Windows which makes it want to run as
elevated, probably because its name contains "update". This makes the
test fail when not running with elevated privileges.

Setting this environment variable tells Windows to ignore the
requirement to run as elevated, allowing the test to pass. This is the
same approach taken in https://reviews.llvm.org/D137772.

Reviewed By: mstorsjo

Differential Revision: https://reviews.llvm.org/D148988
---
 compiler-rt/test/asan/lit.cfg.py | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/compiler-rt/test/asan/lit.cfg.py b/compiler-rt/test/asan/lit.cfg.py
index e8c96f9b58d0..93ed972ccafd 100644
--- a/compiler-rt/test/asan/lit.cfg.py
+++ b/compiler-rt/test/asan/lit.cfg.py
@@ -259,3 +259,12 @@ if not config.parallelism_group:
 
 if config.host_os == 'NetBSD':
   config.substitutions.insert(0, ('%run', config.netbsd_noaslr_prefix))
+
+# Avoid Windows heuristics which try to detect potential installer programs
+# (which may need to run with elevated privileges) and ask if the user wants
+# to run them in that way. This heuristic may match for executables containing
+# the word "update", e.g. `TestCases/asan_update_allocation.cpp`. Set an
+# environment variable indicating that we want to execute them with the current
+# user.
+if config.host_os == 'Windows':
+    config.environment['__COMPAT_LAYER'] = 'RunAsInvoker'
-- 
2.34.1

